proposals/shared-everything-threads/Overview.md

```diff
@@ -222,8 +222,9 @@ An element segment is valid as shared only if its reftype is valid as shared. In
 refer to element segments (e.g. `elem.drop`) are only valid as shared if their element segment is
 shared.
 
-> Note: We could alternatively make an element segment shared iff its elements refer to shared data.
-> It's unclear whether there is a use case for a non-shared element segment containing shared data.
+> Note: We could alternatively make an element segment shared iff its elements refer to shared data,
+> however, we have chosen to add an explicit annotation to avoid future complications if we ever
+> want to add first-class references to element segments.
 
 #### Data Segments
 
```

Comment:
Worth putting a note about whether a shared annotation on data segments is necessary?

Reply:
Should we just go ahead and say in the text that all data segments are shareable without annotation?

Reply:
I'd be on board. In fact, now that I think about it, we already allow `memory.fill` on shared memories, so I'm not sure that we'd ever want to interpret existing data segments as unshared only. Is there another instruction this would be relevant for that I'm forgetting?

Reply:
No, there shouldn't be any semantic issues, just implementation considerations. It would be bad if we paid a performance cost for making all element segments shareable. But we can leave that to implementer feedback and spec that they are all shared for now.

Reply:
> It would be bad if we paid a performance cost for making all element segments shareable. 

Just to double check, I'm only agreeing that we shouldn't add a shared annotation to data segments (bytes). In the absence of any way to manipulate first-class data segments, or other ways to reference data segments, this doesn't seem to have any semantic implications as you say.

I still think we should have a shared annotation on element segments (refs).

Reply:
Ah wait, there are some minor GC implications right? An implementation could hypothetically try to GC a segment that isn't referenced any more, and allowing all data segments to be referenced by shared functions would mean that these data segments could be kept alive cross-thread.

Maybe this is a minor argument in favour of keeping the shared annotation on data segments. Sorry for flip-flopping.

Reply:
Considering that we have data.drop, I think it can matter to implementations whether a segment is shared. That and the uniformity suggests that all segments should conservatively have a shared annotation.

Reply:
Yes, oops, I meant data segments not element segments above. And yes, there are GC implementations because shareable data segments would have to live in the shared space. I'll add a note that we could choose to change this but we'll keep the annotation for now.